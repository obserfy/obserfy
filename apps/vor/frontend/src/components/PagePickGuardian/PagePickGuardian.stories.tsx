import { FC } from "react"
import PagePickGuardian from "./PagePickGuardian"

export default {
  title: "Core/PagePickGuardian",
  component: PagePickGuardian,
  parameters: {
    componentSubtitle: "Just a simple PagePickGuardian",
  },
}

export const Basic: FC = () => <PagePickGuardian />
