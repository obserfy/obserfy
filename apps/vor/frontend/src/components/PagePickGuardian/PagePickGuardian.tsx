import { FC, useState } from "react"
import { Box, Button, Flex, Card } from "theme-ui"
import { Trans, t } from "@lingui/macro"
import { useLingui } from "@lingui/react"
import { navigate } from "../Link/Link"
import BackNavigation from "../BackNavigation/BackNavigation"
import { NEW_STUDENT_URL } from "../../routes"
import Input from "../Input/Input"
import { useNewStudentFormContext } from "../PageNewStudent/NewStudentFormContext"
import Select from "../Select/Select"
import { GuardianRelationship } from "../../hooks/api/students/usePostNewStudent"
import Typography from "../Typography/Typography"
import { ReactComponent as PlusIcon } from "../../icons/plus.svg"

import Icon from "../Icon/Icon"

import { usePostNewGuardian } from "../../hooks/api/guardians/usePostNewGuardian"
import LoadingIndicator from "../LoadingIndicator/LoadingIndicator"
import {
  Guardians,
  useGetSchoolGuardians,
} from "../../hooks/api/guardians/useGetSchoolGuardians"
import TextArea from "../TextArea/TextArea"
import GuardianRelationshipPickerDialog from "../GuardianRelationshipPickerDialog/GuardianRelationshipPickerDialog"

export const PagePickGuardian: FC = () => {
  const { i18n } = useLingui()
  const guardians = useGetSchoolGuardians()
  const { setState } = useNewStudentFormContext()

  const [name, setName] = useState("")
  const [email, setEmail] = useState("")
  const [phone, setPhone] = useState("")
  const [note, setNote] = useState("")
  const [relationship, setRelationship] = useState(GuardianRelationship.Other)
  const [createNew, setCreateNew] = useState(false)
  const { mutateAsync, isLoading } = usePostNewGuardian()

  return (
    <Box sx={{ maxWidth: "maxWidth.sm" }} mx="auto" pb={4}>
      <BackNavigation to={NEW_STUDENT_URL} text="New Student" />
      <Box m={3}>
        <Input
          value={name}
          mb={2}
          label="Guardian Name"
          sx={{ width: "100%" }}
          onChange={(e) => setName(e.target.value)}
        />
        {createNew && (
          <>
            <Select
              label="Relationship"
              mb={2}
              onChange={(e) => setRelationship(parseInt(e.target.value, 10))}
              value={relationship}
            >
              <option value={GuardianRelationship.Other}>Other</option>
              <option value={GuardianRelationship.Mother}>Mother</option>
              <option value={GuardianRelationship.Father}>Father</option>
            </Select>
            <Input
              type="email"
              value={email}
              mb={2}
              label="Email"
              sx={{ width: "100%" }}
              onChange={(event) => setEmail(event.target.value)}
            />
            <Input
              type="phone"
              value={phone}
              mb={3}
              label="Phone"
              sx={{ width: "100%" }}
              onChange={(event) => setPhone(event.target.value)}
            />
            <TextArea
              mb={3}
              label="Note"
              value={note}
              onChange={(e) => setNote(e.target.value)}
            />
            <Flex>
              <Button
                ml="auto"
                variant="outline"
                mr={2}
                onClick={() => setCreateNew(false)}
              >
                <Trans>Cancel</Trans>
              </Button>
              <Button
                data-cy="save-guardian"
                disabled={name === ""}
                onClick={async () => {
                  const payload = { email, name, phone, note }
                  try {
                    const result = await mutateAsync(payload)
                    const resultJson = await result.json()
                    setState((draft) => {
                      draft.guardians.push({
                        id: resultJson.id,
                        relationship,
                      })
                    })
                    await navigate(NEW_STUDENT_URL, {
                      state: { preserveScroll: true },
                    })
                  } catch (e) {
                    Sentry.captureException(e)
                  }
                }}
              >
                {isLoading && <LoadingIndicator color="onPrimary" />}
                <Trans>Save </Trans>
              </Button>
            </Flex>
          </>
        )}
      </Box>
      {!createNew && (
        <>
          <Typography.Body mx={3} mb={2} mt={4} color="textMediumEmphasis">
            <Trans>Select a guardian or create one</Trans>
          </Typography.Body>
          {guardians.data
            ?.filter((guardian) => {
              return guardian.name
                .toLowerCase()
                .includes(name.toLocaleLowerCase())
            })
            ?.map((guardian) => (
              <GuardianCard guardian={guardian} />
            ))}
          <Card
            p={3}
            mx={[0, 3]}
            onClick={() => setCreateNew(true)}
            sx={{
              borderRadius: [0, "default"],
              display: "flex",
              alignItems: "flex-start",
              cursor: "pointer",
              "&:hover": {
                backgroundColor: "primaryLightest",
              },
            }}
          >
            <Icon as={PlusIcon} mt="5px" mr={3} fill="primary" />
            <Typography.Body>
              <Trans>Create</Trans> {name || i18n._(t`new guardian`)}
            </Typography.Body>
          </Card>
        </>
      )}
    </Box>
  )
}

const GuardianCard: FC<{ guardian: Guardians }> = ({ guardian }) => {
  const { setState } = useNewStudentFormContext()
  const [showGuardianSelector, setShowGuardianSelector] = useState(false)

  return (
    <>
      <Card
        onClick={() => setShowGuardianSelector(true)}
        p={3}
        mx={[0, 3]}
        mb={2}
        sx={{
          borderRadius: [0, "default"],
          alignItems: "flex-start",
          display: "flex",
          cursor: "pointer",
          "&:hover": {
            backgroundColor: "primaryLightest",
          },
        }}
      >
        <Typography.Body>{guardian.name}</Typography.Body>
      </Card>
      {showGuardianSelector && (
        <GuardianRelationshipPickerDialog
          onDismiss={() => setShowGuardianSelector(false)}
          onAccept={(relationship) => {
            setState((draft) => {
              draft.guardians.push({
                id: guardian.id,
                relationship,
              })
            })
            navigate(NEW_STUDENT_URL, {
              state: { preserveScroll: true },
            })
          }}
        />
      )}
    </>
  )
}

export default PagePickGuardian
