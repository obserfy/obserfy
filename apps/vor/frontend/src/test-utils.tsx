/* eslint-disable import/no-extraneous-dependencies */
import { FC, ReactElement } from "react"
import { QueryClient, QueryClientProvider } from "react-query"
import { ThemeProvider, Box, Theme } from "theme-ui"
import { render, RenderOptions, RenderResult } from "@testing-library/react"
import {
  createHistory,
  createMemorySource,
  LocationProvider,
} from "@reach/router"
import { I18nProvider } from "@lingui/react"
import { i18n } from "@lingui/core"
import { en } from "make-plural/plurals"
import theme from "./gatsby-plugin-theme-ui"
// @ts-ignore
import enCatalog from "../i18n/lingui/en/messages.js"

// any used here because the catalogs are generated b lingui
// not our responsibility
i18n.loadLocaleData("en", { plurals: en })
i18n.load("en", enCatalog.messages as any)
i18n.activate("en")

const queryClient = new QueryClient()
const history = createHistory(createMemorySource("/"))

const AllProviders: FC = ({ children }) => {
  return (
    <I18nProvider i18n={i18n}>
      <QueryClientProvider client={queryClient}>
        <LocationProvider history={history}>
          <ThemeProvider theme={theme as Theme}>
            <Box sx={{ fontSize: [16, 20] }}>{children}</Box>
          </ThemeProvider>
        </LocationProvider>
      </QueryClientProvider>
    </I18nProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult => render(ui, { wrapper: AllProviders, ...options })

// re-export everything
export * from "@testing-library/react"

// override render method
export { customRender as render }
