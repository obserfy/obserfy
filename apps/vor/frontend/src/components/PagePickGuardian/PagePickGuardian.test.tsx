import { render } from "../../test-utils"
import { Basic } from "./PagePickGuardian.stories"

describe("PagePickGuardian", () => {
  it("should render correctly", () => {
    const { container } = render(<Basic />)
    expect(container).toMatchSnapshot()
  })
})
