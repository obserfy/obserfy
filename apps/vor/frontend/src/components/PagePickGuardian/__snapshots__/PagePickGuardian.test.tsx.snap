// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PagePickGuardian should render correctly 1`] = `
.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-size: 16px;
}

@media screen and (min-width: 40em) {
  .emotion-0 {
    font-size: 20px;
  }
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-left: auto;
  margin-right: auto;
  padding-bottom: 32px;
  max-width: 640px;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-left: -8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  max-width: 100%;
  height: auto;
  margin: 16px;
  margin-right: 4px;
  fill: var(--theme-ui-colors-icon);
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 24px;
  height: 24px;
  fill: var(--theme-ui-colors-textMediumEmphasis);
}

.emotion-4 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: var(--theme-ui-colors-text);
  font-size: 14px;
  text-rendering: optimizeLegibility;
  line-height: 1.5;
  margin-bottom: 0;
  color: var(--theme-ui-colors-textMediumEmphasis);
}

.emotion-5 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin: 16px;
}

.emotion-6 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: var(--theme-ui-colors-textMediumEmphasis);
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-size: 14px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: var(--theme-ui-colors-textMediumEmphasis);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-7 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  padding-bottom: 4px;
}

.emotion-8 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-9 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: block;
  width: 100%;
  padding: 8px;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  font-size: inherit;
  line-height: inherit;
  border: 1px solid;
  border-radius: 4px;
  color: inherit;
  background-color: transparent;
  border-radius: 6px;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  width: auto;
  border-color: var(--theme-ui-colors-border);
  background-color: var(--theme-ui-colors-surface);
  font-size: 16px;
  padding: 12px;
  color: var(--theme-ui-colors-text);
  margin-bottom: 8px;
  --theme-ui-input-autofill-bg: var(--theme-ui-colors-background);
  width: 100%;
  width: 100%;
}

.emotion-9:autofill,
.emotion-9:autofill:hover,
.emotion-9:autofill:focus {
  box-shadow: inset 0 0 0 1000px var(--theme-ui-input-autofill-bg);
  font-size: inherit;
}

.emotion-9:autofill:first-line,
.emotion-9:autofill:hover:first-line,
.emotion-9:autofill:focus:first-line {
  font-size: 1rem;
}

.emotion-9:-webkit-autofill,
.emotion-9:-webkit-autofill:hover,
.emotion-9:-webkit-autofill:focus {
  box-shadow: inset 0 0 0 1000px var(--theme-ui-input-autofill-bg);
  font-size: inherit;
}

.emotion-9:-webkit-autofill:first-line,
.emotion-9:-webkit-autofill:hover:first-line,
.emotion-9:-webkit-autofill:focus:first-line {
  font-size: 1rem;
}

.emotion-9:hover,
.emotion-9:focus {
  border-color: var(--theme-ui-colors-primary);
}

.emotion-9::-webkit-input-placeholder {
  font-style: italic;
}

.emotion-9::-moz-placeholder {
  font-style: italic;
}

.emotion-9:-ms-input-placeholder {
  font-style: italic;
}

.emotion-9::placeholder {
  font-style: italic;
}

.emotion-9:disabled {
  opacity: 0.68;
  border-color: var(--theme-ui-colors-border);
  color: var(--theme-ui-colors-textDisabled);
}

.emotion-10 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: var(--theme-ui-colors-text);
  font-size: 14px;
  text-rendering: optimizeLegibility;
  line-height: 1.5;
  margin-left: 16px;
  margin-right: 16px;
  margin-bottom: 8px;
  margin-top: 32px;
  color: var(--theme-ui-colors-textMediumEmphasis);
}

.emotion-11 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  background-color: var(--theme-ui-colors-surface);
  border-radius: 6px;
  box-shadow: rgba(0, 0, 0, 0.1) 0px 0px 1px,rgba(0, 0, 0, 0.27) 0px 2px 4px -2px;
  padding: 16px;
  margin-left: 0;
  margin-right: 0;
  border-radius: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  cursor: pointer;
}

@media screen and (min-width: 40em) {
  .emotion-11 {
    margin-left: 16px;
    margin-right: 16px;
  }
}

@media screen and (min-width: 40em) {
  .emotion-11 {
    border-radius: 6px;
  }
}

.emotion-11:hover {
  background-color: var(--theme-ui-colors-primaryLightest);
}

.emotion-12 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  max-width: 100%;
  height: auto;
  margin-top: 5px;
  margin-right: 16px;
  fill: var(--theme-ui-colors-primary);
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 20px;
  height: 20px;
}

.emotion-13 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: var(--theme-ui-colors-text);
  font-size: 14px;
  text-rendering: optimizeLegibility;
  line-height: 1.5;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <a
        href="/dashboard/students/new"
        style="display: block;"
      >
        <div
          class="emotion-2"
        >
          <div
            class="emotion-3"
          />
          <p
            class="emotion-4"
          >
            New Student
          </p>
        </div>
      </a>
      <div
        class="emotion-5"
      >
        <label
          class="emotion-6"
        >
          <div
            class="emotion-7"
          >
            Guardian Name
          </div>
          <div
            class="emotion-8"
          >
            <input
              class="emotion-9"
              value=""
            />
          </div>
        </label>
      </div>
      <p
        class="emotion-10"
      >
        Select a guardian or create one
      </p>
      <div
        class="emotion-11"
      >
        <div
          class="emotion-12"
        />
        <p
          class="emotion-13"
        >
          Create
           
          new guardian
        </p>
      </div>
    </div>
  </div>
</div>
`;
